### PR TITLE
Improve frozen abi dx a bit and document it

### PR DESCRIPTION
**Pull request by ryoqun**
#### Problem

Frozen abi needs some love periodically.

#### Summary of Changes

It's another time again. This pr is cumulative one containing multiple changes relating to frozen abi.

See individual commits.

### CODE REVIEW

**Comment by apfitzge** [comment]
I assume this is a response to my discord comment here: https://discord.com/channels/428295358100013066/439194979856809985/1261428595814039603

While these changes fix it for my specific case of changing the hash function, I am still curious about the larger question of why the Bank needs a frozen Abi at all.
I understand why some other structs in accounts/runtime do. But Bank is the one that is most annoying, given its current state as a "god" class.

**Comment by ryoqun** [comment]
> While these changes fix it for my specific case of changing the hash function, I am still curious about the larger question of why the Bank needs a frozen Abi at all.

hehe, thanks for jumping in. i was too early to mark this pr out of draft.. I pushed a commit for the explanation. is there specific questions not answered by this still?: https://github.com/anza-xyz/agave/pull/2131/commits/99cdd64a18938d25ad7e99e345e00056043c4b63

**Comment by apfitzge** [comment]
> > While these changes fix it for my specific case of changing the hash function, I am still curious about the larger question of why the Bank needs a frozen Abi at all.
> 
> hehe, thanks for jumping in. i was too early to mark this pr out of draft.. I pushed a commit for the explanation. is there specific questions not answered by this still?: [99cdd64](https://github.com/anza-xyz/agave/commit/99cdd64a18938d25ad7e99e345e00056043c4b63)

Had responded on discord since you commented there.

But this explanation, to me, seems to justify that its' necessary because of the test structure...which we could also change to make that **not** required.

**Comment by ryoqun** [comment]
> But this explanation, to me, seems to justify that its' necessary because of the test structure...which we could also change to make that **not** required.

yeah. as written in the commit, i was reluctant to copy all the type arguments of `serialize_bank_snapshot_with`. A good news is that i found a good compromise: https://github.com/anza-xyz/agave/pull/2131/commits/d9615f4a038841707b4beb69e8a36bd2ba2bdf17

File: frozen-abi/src/abi_example.rs [hunk 1]
```diff
@@ -431,6 +431,28 @@ impl<
     }
 }
 
+#[cfg(not(target_os = "solana"))]
+impl<T: Clone + std::cmp::Eq + std::hash::Hash + AbiExample, S: Clone + AbiExample> AbiExample
```
Comment (apfitzge) [inline frozen-abi/src/abi_example.rs]: given the other changes, think we can drop adding this.

Comment (ryoqun) [reply frozen-abi/src/abi_example.rs]: done by dropping the commit from the pr.

**Comment by yihau** [comment]
looks good to me but I would like to wait for @apfitzge to have a chance to take a look!

**Review by apfitzge — APPROVED** [review]

**Pull request merged**